LIT-2504 - Implement scheduled tasks - Capacity NFT top-up

MaximusHaximus opened the pull request:
Adds a scheduled task runner leveraging Agenda for task scheduling + execution - we can set an interval in every X hours/days/whathaveyou, or use CRON syntax to set specific day/times.

Defines a single job, `auto-top-up`, which will be responsible for minting capacity NFTs and transferring them to specific target addresses.

- Uses `unbuild` for builds
- `pnpm` for packag management
- Mocha & Chai for testing
- .nvmrc is committed; recommend using nvm as it will use the correct NodeJS version (currently v18)
- Awaity to process 3 recipients concurrently (middleground for spammy vs. takes too long) - 😿 concurrent nonces will be a problem, made it run serially for now.
- VError for errors w/ context

MaximusHaximus commented on packages/lit-task-auto-top-up/.mocharc.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "require": ["tsx"],
```
TSX does on-the-fly transpilation so you can debug in IDE without buildng.

MaximusHaximus commented on packages/lit-task-auto-top-up/package.json:
```diff
@@ -0,0 +1,46 @@
+  "description": "",
+  "type": "module",
+  "packageManager": "pnpm@8.15.5",
+  "exports": {
```
Unbuild auto-detects output formats based on these keys

MaximusHaximus commented on packages/lit-task-auto-top-up/src/taskHandler.ts:
```diff
@@ -0,0 +1,56 @@
+
+        return settledResult;
+      },
+      3 // Let us mint 3 at a time so this isn't too slow but isn't a spammy deluge
```
@Ansonhkg @FedericoAmura I have a suspicion this concurrency might actually fail due to `nonce` being re-used if multiple requests are sent for the same account concurrently. Can you confirm if that's the case?

Ansonhkg replied:
Yes, this will fail - that's why in the current e2e tests, they are all in async rather than parallel. A previous solution involved queueing operations to prevent this issue - by creating a wrapper function that interacts with a locally hosted API server before the test. This server manages a queue in memory. When the API endpoint is called, it doesn't return a success response (true) until it has completed the previous operation.

MaximusHaximus replied:
OK, modified it to use `mapSeries` instead.

MaximusHaximus commented on packages/lit-task-auto-top-up/src/types/types.ts:
```diff
@@ -0,0 +1,10 @@
+import { z } from 'zod';
```
Zod <-> Rust inter-op...might be worth looking into for single-source-of-truth with the node Rust codebase

https://www.reddit.com/r/rust/comments/17dyjyd/i_made_a_crate_for_bridging_types_between_rust/
https://github.com/spencerkohan/type_reflect

FedericoAmura commented on packages/lit-task-scheduler/src/initializeTasks.ts:
```diff
@@ -0,0 +1,14 @@
+const AUTO_TOP_UP_TASK_NAME = `daily ${autoTopUpName}`;
+export default async function initializeTasks(taskClient: Agenda) {
+  // Idempotent calls here only. Yay. :)
+  await taskClient.every('1 day', AUTO_TOP_UP_TASK_NAME);
```
Running the task like this won't tie the execution time with the time we start this app? Do we want that?

MaximusHaximus replied:
Good catch - I was meaning to set this but forgot.  Fixed; now it will run at 10PM GMT, a couple of hours before any existing NFT will expire since they always expire at midnight UTC.  WDYT?

FedericoAmura commented on packages/lit-task-auto-top-up/src/taskHandler.ts:
```diff
@@ -0,0 +1,55 @@
+    const results = await mapSeries<RecipientDetail, PromiseSettledResult<TaskResult>>(
+      recipientList,
+      async (recipientDetail) => {
+        const [settledResult] = await Promise.allSettled([handleRecipient({ recipientDetail })]);
```
Cool trick

FedericoAmura commented on packages/lit-task-auto-top-up/src/taskHelpers.ts:
```diff
@@ -0,0 +1,46 @@
+      consola.error(error, JSON.stringify(VError.info(error), null, 2));
+    });
+
+    if (successes.length > 1 && successes.length === errors.length) {
```
```suggestion
    if (results.length === errors.length) {
```
I guess that was the intent based on the comment

MaximusHaximus replied:
🤦 you're right :) Fixed.

MaximusHaximus commented on packages/lit-task-auto-top-up/src/actions/getRecipientList.ts:
```diff
@@ -0,0 +1,16 @@
+const recipientListInstance = new RecipientList(getConfig());
+
+export default async function getRecipients(): Promise<RecipientDetail[]> {
+  return recipientListInstance.fetchRecipientList();
```
Good catch 🤦  Webstorm was dutifully marking the method as unused but for some reason I ignored it 😢  Fixed.

I was torn about this too, because I tend to lean towards 'fail fast and hard' thinking for similar reasons.

My thinking was that if someone was unlikely to see an alert about the list being invalid, then they would miss the alert about the entire service being down, and it would be better to still keep minting for the existing list than to have _everyone_ no longer get tokens... if it takes some time to fix the source, then at least other consumers are being served.

Mainly, what tipped me into implementing this was considering that even if we _do_ have perfect monitoring of the service, _if the file is invalid_, that monitoring won't trigger  and let people who can fix it know_ until the job fails at least once_.  At that point, _we've missed minting for some group of users_, so end-users will be impacted by the error in file format.

This way, we can alert that the source-of-truth is bad without the rest of the previously defined recipients missing a mint.  WDYT?

Also, I'm going to add a commit hook and PR check that validates the JSON matches our schema... but I just know someone will edit it directly in Github and bypass them both with some invalid input sometime.... :P :)

MaximusHaximus replied:
Added commit hook and PR checks for the JSON file, and updated the schema to validate the recipient addresses using strict validation provided by `Viem` 🎉 

Given how unlikely it will be for the file to be malformed, given the repository configuration would require that someone _disable branch protection on the repo_, then _use the github web UI to edit `main` directly_ in order to make the file actually end up in `main` in an invalid shape, I think this functionality is now really only about being resilient to transient issues loading the file from `Github`, so I'm going to leave the local cache in for now, unless you have a very strong opinion against it :D

MaximusHaximus commented:
@FedericoAmura Added git hook and CI check that validates the JSON :)
Also cleaned up boundaries between the worker, the task client and the task definition, and normalized Config between both packages and moved some stuff around to be clearer.

Let me know what you think. :+1: 💖

MaximusHaximus commented:
@FedericoAmura Added validation for recipientAddress that uses strict address validation (including checksum checks!) from `Viem` :).  Check out the `.refine() usage in `Zod`: https://github.com/LIT-Protocol/lit-scheduled-tasks/pull/1/commits/b68fc554f9d58929173cfa9e64dd56715a30a524

It's fly like superman ✈

FedericoAmura commented on README.md:
````diff
@@ -0,0 +1,48 @@
+
+#### To prepare for either production usage or local development, you must install dependencies and build the packages.
+
+```zsh
````
Just curious, why `zsh` and not `bash`?

MaximusHaximus replied:
Webstorm provide super helpful little 'run' buttons when viewing this stuff in the IDE, and it uses the shell provided to run it... and zsh is the most popular these days so I'm just betting on commonality

FedericoAmura commented on packages/lit-task-client/src/singletons/getConfig.ts:
```diff
@@ -0,0 +1,21 @@
+/**
+ * `getConfig`returns singleton instance of the Config class, which is initialized at module-scope
+ * when this module is loaded. WARNING: Loading this module may cause an immediate exception, as it
```
Not anymore

MaximusHaximus replied:
Fixed :)

MaximusHaximus commented:
Thanks for the thoughtful reviews @FedericoAmura 💖
